src/internal/observable/combineLatest.ts

```diff
@@ -188,117 +158,88 @@ export function combineLatest<T extends Record<string, ObservableInput<any>>>(
  * @see {@link merge}
  * @see {@link withLatestFrom}
  *
- * @param {ObservableInput} [observables] An array of input Observables to combine with each other.
+ * @param sources An array of input Observables to combine with each other.
  * An array of Observables must be given as the first argument.
- * @param {function} [project] An optional function to project the values from
+ * @param project An optional function to project the values from
  * the combined latest values into a new value on the output Observable.
- * @param {SchedulerLike} [scheduler=null] The {@link SchedulerLike} to use for subscribing to
- * each input Observable.
- * @return {Observable} An Observable of projected values from the most recent
+ * @return An Observable of projected values from the most recent
  * values from each input Observable, or an array of the most recent values from
  * each input Observable.
  */
-export function combineLatest<O extends ObservableInput<any>, R>(...args: any[]): Observable<R> | Observable<ObservedValueOf<O>[]> {
-  const scheduler = popScheduler(args);
-  const resultSelector = popResultSelector(args);
+export function combineLatest<O extends ObservableInput<any>, R>(
+  sources: Record<string, O> | O[],
+  resultSelector?: (...value: any[]) => R
+): Observable<R> | Observable<ObservedValueOf<O>[]> {
+  const parts = arrayOrObject(sources);
 
-  const { args: observables, keys } = argsArgArrayOrObject(args);
+  if (!parts) {
+    throw new TypeError('sources must be an array or object');
+  }
+
+  const { args: observables, keys } = parts;
```

Comment:
I propose to separate `arrayOrObject` and `argsArgArrayOrObject`, make a check "sources must be an array or object" inside `arrayOrObject`. This will make it easier for us if/when we want to make similar changes to `zip` and `forkJoin`.

Reply:
Let's do this in a follow up. I need to think about the best approach a little bit. Incremental refactors.

Every change to thrown errors is a change to our external API, and therefore a breaking change.

for example, if someone is doing this, and we change the error message to be more specific, we'll break them:

```ts
try {
  combineLatest(someVariable)
} catch (err) {
  if (err.message === 'sources must be an array or object') {
    // do something.
  }
}
```

I'm not saying we shouldn't make the change, just that we should make sure we're moving deliberatly.